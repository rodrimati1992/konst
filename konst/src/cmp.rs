//! Comparisong-related items.

mod cmp_wrapper;

mod const_cmp;

pub use self::{
    cmp_wrapper::CmpWrapper,
    const_cmp::{ConstCmp, ConstCmpUnref, IsAConstCmp},
};

#[doc(no_inline)]
pub use crate::polymorphism::kinds::{IsNotStdKind, IsRefKind, IsStdKind};
