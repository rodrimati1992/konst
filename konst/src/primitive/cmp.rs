use core::cmp::Ordering;

macro_rules! declare_int_cmp_fn {
    (
        ($type:ty, ($cmp_fn_name:ident))
        docs( $docs_eq:expr, $docs_cmp:expr, )
    ) => {
        crate::__delegate_const_ord! {
            #[doc = $docs_cmp]
            pub const fn $cmp_fn_name(copy left: $type, right: $type) -> Ordering {
                cmp_int!(left, right)
            }
        }
    };
}

__declare_fns_with_docs! {
    (u8, (cmp_u8))
    (u16, (cmp_u16))
    (u32, (cmp_u32))
    (u64, (cmp_u64))
    (u128, (cmp_u128))
    (usize, (cmp_usize))

    (i8, (cmp_i8))
    (i16, (cmp_i16))
    (i32, (cmp_i32))
    (i64, (cmp_i64))
    (i128, (cmp_i128))
    (isize, (cmp_isize))

    (bool, (cmp_bool))
    (char, (cmp_char))

    docs(default)

    macro = declare_int_cmp_fn!(),
}

__declare_fns_with_docs! {
    (Option<u8>, (eq_option_u8, cmp_option_u8))
    (Option<i8>, (eq_option_i8, cmp_option_i8))
    (Option<u16>, (eq_option_u16, cmp_option_u16))
    (Option<i16>, (eq_option_i16, cmp_option_i16))
    (Option<u32>, (eq_option_u32, cmp_option_u32))
    (Option<i32>, (eq_option_i32, cmp_option_i32))
    (Option<u64>, (eq_option_u64, cmp_option_u64))
    (Option<i64>, (eq_option_i64, cmp_option_i64))
    (Option<u128>, (eq_option_u128, cmp_option_u128))
    (Option<i128>, (eq_option_i128, cmp_option_i128))
    (Option<usize>, (eq_option_usize, cmp_option_usize))
    (Option<isize>, (eq_option_isize, cmp_option_isize))
    (Option<bool>, (eq_option_bool, cmp_option_bool))
    (Option<char>, (eq_option_char, cmp_option_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        params(l, r)
        eq_comparison = l == r,
        cmp_comparison = cmp_int!(l, r),
        parameter_copyability = copy,
    ),
}
