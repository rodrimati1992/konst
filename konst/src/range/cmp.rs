use core::ops::{Range, RangeInclusive};

use crate::cmp::{CmpWrapper, ConstCmp, IsAConstCmp, IsStdKind};

macro_rules! shared_range_impls {
    ($type:ty, $eq_fn_name:ident) => {
        impl ConstCmp for $type {
            type Kind = IsStdKind;
        }
        impl CmpWrapper<$type> {
            /// Compares `self` and `other` for equality.
            #[inline(always)]
            pub const fn const_eq(&self, other: &$type) -> bool {
                $eq_fn_name(&self.0, other)
            }
        }
    };
}

macro_rules! declare_range_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(Range {
                    start: range.start,
                    end: range.end,
                })
            }
        }

        __delegate_const_eq! {
            skip_coerce;
            for['a,]

            #[doc = $docs_eq]
            pub const fn $eq_fn_name(ref left: &'a $type, right: &'a $type) -> bool {
                left.start == right.start && left.end == right.end
            }
        }
    };
}

__declare_fns_with_docs! {
    (Range<u8>,    (eq_range_u8))
    (Range<u16>,   (eq_range_u16))
    (Range<u32>,   (eq_range_u32))
    (Range<u64>,   (eq_range_u64))
    (Range<u128>,  (eq_range_u128))
    (Range<usize>, (eq_range_usize))

    (Range<char>, (eq_range_char))

    docs(default)

    macro = declare_range_cmp_fns!(),
}

macro_rules! declare_rangeinclusive_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(RangeInclusive::new(*range.start(), *range.end()))
            }
        }

        #[doc = $docs_eq]
        pub const fn $eq_fn_name(left: &$type, right: &$type) -> bool {
            *left.start() == *right.start() && *left.end() == *right.end()
        }
    };
}

__declare_fns_with_docs! {
    (RangeInclusive<u8>,    (eq_rangeinc_u8,))
    (RangeInclusive<u16>,   (eq_rangeinc_u16))
    (RangeInclusive<u32>,   (eq_rangeinc_u32))
    (RangeInclusive<u64>,   (eq_rangeinc_u64))
    (RangeInclusive<u128>,  (eq_rangeinc_u128))
    (RangeInclusive<usize>, (eq_rangeinc_usize))

    (RangeInclusive<char>, (eq_rangeinc_char))

    docs(default)

    macro = declare_rangeinclusive_cmp_fns!(),
}
