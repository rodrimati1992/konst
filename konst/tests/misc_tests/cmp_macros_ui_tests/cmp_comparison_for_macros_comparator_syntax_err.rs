use konst::cmp;

use std::cmp::Ordering as Orde;

const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8] asdasodijasd);
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8] asdasodijasd);
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8] asdasodijasd);
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8] asdasodijasd);
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8] asdasodijasd);
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8] asdasodijasd);
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8] asdasodijasd);

// type annotations aren't supported
const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x: &u32| 0u8);


const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x| );
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x| );
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x| );
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x| );
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x| );
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x| );
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x| );

// type annotations aren't supported
const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);


const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x, y| );
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x, y| );
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x, y| );
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x, y| );
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x, y| );
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x, y| );
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x, y| );

const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], -);
const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], -);
const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], -);
const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], -);
const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], -);
const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], -);
const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], -);

fn main() {}