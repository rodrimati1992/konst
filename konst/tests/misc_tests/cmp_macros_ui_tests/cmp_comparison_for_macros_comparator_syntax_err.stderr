error: no rules expected `asdasodijasd`
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:5:56
  |
5 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8] asdasodijasd);
  |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
  |                                                       |
  |                                                       help: missing comma here
  |
  = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:6:56
  |
6 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8] asdasodijasd);
  |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
  |                                                       |
  |                                                       help: missing comma here
  |
  = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:7:57
  |
7 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8] asdasodijasd);
  |                                                        -^^^^^^^^^^^^ no rules expected this token in macro call
  |                                                        |
  |                                                        help: missing comma here
  |
  = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:8:56
  |
8 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8] asdasodijasd);
  |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
  |                                                       |
  |                                                       help: missing comma here
  |
  = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:9:56
  |
9 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8] asdasodijasd);
  |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
  |                                                       |
  |                                                       help: missing comma here
  |
  = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:10:56
   |
10 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8] asdasodijasd);
   |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
   |                                                       |
   |                                                       help: missing comma here
   |
   = note: while trying to match sequence start

error: no rules expected `asdasodijasd`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:11:56
   |
11 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8] asdasodijasd);
   |                                                       -^^^^^^^^^^^^ no rules expected this token in macro call
   |                                                       |
   |                                                       help: missing comma here
   |
   = note: while trying to match sequence start

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:14:59
   |
14 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:15:59
   |
15 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:16:60
   |
16 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                            ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:17:59
   |
17 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:18:59
   |
18 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:19:59
   |
19 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:20:59
   |
20 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:23:60
   |
23 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:24:60
   |
24 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:25:61
   |
25 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x| );
   |                                                             ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:26:60
   |
26 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:27:60
   |
27 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:28:60
   |
28 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:29:60
   |
29 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x| );
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$key_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                              ^^^^^^^^^^^^^^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:32:59
   |
32 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:33:59
   |
33 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:34:60
   |
34 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                            ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:35:59
   |
35 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:36:59
   |
36 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:37:59
   |
37 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: no rules expected `:`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:38:59
   |
38 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x: &u32, y: &u32| 0u8);
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param| $key_expr:expr $(,)*) => {{
   |                                            ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:41:63
   |
41 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:42:63
   |
42 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:43:64
   |
43 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], |x, y| );
   |                                                                ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:44:63
   |
44 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:45:63
   |
45 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:46:63
   |
46 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:47:63
   |
47 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], |x, y| );
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$eq_expr:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, |$l:pat_param, $r:pat_param| $eq_expr:expr $(,)*) => {{
   |                                                            ^^^^^^^^^^^^^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:49:57
   |
49 | const _: bool = cmp::const_eq_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:50:57
   |
50 | const _: bool = cmp::const_ne_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_eq_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:51:58
   |
51 | const _: Orde = cmp::const_cmp_for!(slice; [0u8], [3u8], -);
   |                                                          ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:52:57
   |
52 | const _: bool = cmp::const_lt_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:53:57
   |
53 | const _: bool = cmp::const_le_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:54:57
   |
54 | const _: bool = cmp::const_gt_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^

error: no rules expected `-`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_comparator_syntax_err.rs:55:57
   |
55 | const _: bool = cmp::const_ge_for!(slice; [0u8], [3u8], -);
   |                                                         ^ no rules expected this token in macro call
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |     ($left:expr, $right:expr, ) => {{
   |                             ^
